### PR TITLE
Added onboarding automation script

### PR DESCRIPTION
**Pull request by amanshanbhag**
This script automates the onboarding process -- one step closer to self service. It guides the user through all the steps from the workshop, asking for details periodically (and confirmations)

### CODE REVIEW

File: 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+
+# Function to setup environment variables
+setup_env_vars() {
+    echo -e "${BLUE}=== Setting Up Environment Variables ===${NC}"
+
+    echo -e "${YELLOW}Downloading configuration script...${NC}"
+    curl 'https://static.us-east-1.prod.workshops.aws/public/4a96b723-a0e8-46c0-a78f-0dc6a37d8f0d/static/scripts/create_config.sh' --output create_config.sh
+    echo -e "${GREEN}✅ Configuration script downloaded${NC}"
+    
+    echo -e "${BLUE}Enter CloudFormation stack name (default: sagemaker-hyperpod):${NC}"
```
Comment (nghtm) [inline 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh:115]: Lets be more descriptive.

Enter the name of the SageMaker VPC CloudFormation stack that was deployed as a prerequisite. (default: sagemaker-hyperpod):

File: 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+    cd awsome-distributed-training/1.architectures/5.sagemaker-hyperpod/LifecycleScripts/
+
+    echo -e "${YELLOW}Did you deploy the optional hyperpod-observability CloudFormation stack? (yes/no)${NC}"
+    read -e DEPLOYED_OBSERVABILITY
+
+    if [ "$DEPLOYED_OBSERVABILITY" == "yes" ]; then
+        echo -e "${BLUE}Enabling observability in LCS...${NC}"
+        sed -i.bak 's/enable_observability = False/enable_observability = True/' base-config/config.py
+        rm base-config/config.py.bak
+        echo -e "${GREEN}✅ Observability enabled in configuration${NC}"
```
Comment (nghtm) [inline 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh:180]: Lets change this to "Lifecycle Scripts modified successfully! Observability enabled in config.py"

File: 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+}
+
+validate_cluster_config() {
+    echo "Validating your cluster configuration..."
+    # TODO: MAKE SURE PACKAGES ARE INSTALLED HERE!!
+
+    curl -O https://raw.githubusercontent.com/aws-samples/awsome-distributed-training/main/1.architectures/5.sagemaker-hyperpod/validate-config.py
+
+    # check config for known issues
+    python3 validate-config.py --cluster-config cluster-config.json --provisioning-parameters provisioning_parameters.json
```
Comment (nghtm) [inline 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh:352]: When I ran through the setup in my local IDE, the script failed with this error:

Now I have to restart whole configuration. Can we give customers a change to 1/ pip install json schema if this error is encountered, 2/ provide warning about json schema requirement prior to running script, or 3/ some better idea...
```
ℹ️  Validating the generated configuration before proceeding
Validating your cluster configuration...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  8363  100  8363    0     0  78847      0 --:--:-- --:--:-- --:--:-- 79647
Traceback (most recent call last):
  File "/Users/nghtm/VScode/pr484/awsome-distributed-training/1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/validate-config.py", line 6, in <module>
    from jsonschema import validate
ModuleNotFoundError: No module named 'jsonschema
```

Comment (nghtm) [inline 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh:352]: I also hit this error when testing, becuase the region my CLI was configured was us-east-1 but I was using the script to deploy cluster in us-west-2. 

```
botocore.exceptions.ClientError: An error occurred (InvalidSubnetID.NotFound) when calling the DescribeSubnets operation: The subnet ID 'subnet-0d36904857b8e3f61' does not exist
```

We can mitigate this error by passing the --region parameter to `validate-config.py` incase the region set by CLI is different from what user passed to the script. 
```
python3 validate-config.py --cluster-config cluster-config.json --provisioning-parameters provisioning_parameters.json --region $AWS_REGION
```

File: 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+        exit 1
+    fi
+}
+
+# Function to setup environment variables
+setup_env_vars() {
+    echo -e "${BLUE}=== Setting Up Environment Variables ===${NC}"
+
+    echo -e "${YELLOW}Downloading configuration script...${NC}"
+    curl 'https://static.us-east-1.prod.workshops.aws/public/4a96b723-a0e8-46c0-a78f-0dc6a37d8f0d/static/scripts/create_config.sh' --output create_config.sh
```
Comment (sean-smith) [inline 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh:112]: Let’s move this script to the GitHub repository

File: 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+    
+    echo -e "${BLUE}Enter CloudFormation stack name (default: sagemaker-hyperpod):${NC}"
+    read -e CF_STACK_NAME
+    CF_STACK_NAME=${CF_STACK_NAME:-sagemaker-hyperpod}
+
+    if [ "$CF_STACK_NAME" != "sagemaker-hyperpod" ]; then
+        # Replace 'sagemaker-hyperpod' with the user-provided stack name in the create_config.sh script    
+        echo -e "${YELLOW}Using custom stack name: ${GREEN}$CF_STACK_NAME${NC}"
+        echo -e "${BLUE}Updating configuration script...${NC}"
+        sed -i.bak "s/\${STACK_ID_VPC:=sagemaker-hyperpod}/\${STACK_ID_VPC:=$CF_STACK_NAME}/" create_config.sh
```
Comment (sean-smith) [inline 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh:123]: Just pass these in as environment variables. No need to replace in file.

File: 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+    read -e DEPLOYED_OBSERVABILITY
+
+    if [ "$DEPLOYED_OBSERVABILITY" == "yes" ]; then
+        echo -e "${BLUE}Enabling observability in LCS...${NC}"
+        sed -i.bak 's/enable_observability = False/enable_observability = True/' base-config/config.py
+        rm base-config/config.py.bak
+        echo -e "${GREEN}✅ Observability enabled in configuration${NC}"
+
+        echo -e "${BLUE}Attaching IAM policies for observability to $ROLENAME${NC}"
+        aws iam attach-role-policy --role-name $ROLENAME --policy-arn arn:aws:iam::aws:policy/AmazonPrometheusRemoteWriteAccess
```
Comment (sean-smith) [inline 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh:183]: Capture errors here if user doesn’t have Admin permissions I.e. a try/catch block

**Review by nghtm — COMMENTED** [review]
Nice work Aman! Comments / Improvements provided below. This will help immensely, invent and simplify!

File: 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh (hunk 3 above)
Comment (amanshanbhag) [reply 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh:352]: Added this in to a "prereq message". I'm a bit hesitant to install packages

File: 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+
+    print_header "🎉 Cluster Creation Script Completed! 🎉"
+
+    # Instructions for next steps
+    echo -e "${GREEN}Congratulations! You've completed all the preparatory steps.${NC}"
+    echo -e "${YELLOW}Next Steps:${NC}"
+    echo -e "${YELLOW}Run the following command to create the cluster. Exiting this script!${NC}"
+
+    # Command to create the cluster
+    echo -e "${YELLOW}aws sagemaker create-cluster \\"
```
Comment (nghtm) [inline 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh:454]: Can you add a space before this command. And make change the color to green to make what needs to be copy pasted more clear...

Screenshot 2024-11-07 at 5 33 29 PM

File: 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+    read -e -p "$prompt [$default]: " input
+    echo "${input:-$default}"    
+}
+
+# Function to write the cluster-config.json file
+create_config() {
+    echo -e "\n${BLUE}=== Lifecycle Scripts Setup Complete ===${NC}"
+
+    # Get controller machine details
+    CONTROLLER_NAME=$(get_input "Enter the name for the controller instance group" "controller-machine")
```
Comment (nghtm) [inline 1.architectures/5.sagemaker-hyperpod/automate-smhp-slurm/automate-cluster-creation.sh:215]: Lets add a login machine, we can prompt this to users as optional, but it is best practice and we have added to the workshop. `m5.4xlarge ` as default

**Review by nghtm — APPROVED** [review]
Approved after comments addressed.

**Pull request merged**